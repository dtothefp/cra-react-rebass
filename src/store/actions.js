import {
  ADD_FILTER,
  REMOVE_FILTER,
  SET_PHARMACIES
} from './constants';

export const setPharmacies = (pharmacies) => ({
  type: SET_PHARMACIES,
  pharmacies
});

export const addFilter = (value) => ({
  type: ADD_FILTER,
  value: value.toUpperCase(),
});

export const removeFilter = (value) => ({
  type: REMOVE_FILTER,
  value: value.toUpperCase(),
});
