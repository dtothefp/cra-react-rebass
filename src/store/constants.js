export const ADD_FILTER = `ADD_FILTER`;
export const REMOVE_FILTER = `REMOVE_FILTER`;
export const SET_PHARMACIES = `SET_PHARMACIES`;
