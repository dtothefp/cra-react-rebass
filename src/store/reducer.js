import {
  ADD_FILTER,
  REMOVE_FILTER,
  SET_PHARMACIES
} from './constants';

export default (state = {}, action) => {
  let updatedState;

  switch (action.type) {
    case SET_PHARMACIES:
      updatedState = {
        ...state,
        pharmacies: action.pharmacies
      };
      break;
    case ADD_FILTER:
      updatedState = {
        ...state,
        filter: [ ...state.filter, action.value ]
      };
      break;
    case REMOVE_FILTER:
     updatedState = {
        ...state,
        filter: state.filter.filter((value) => value !== action.value),
      };
      break;
    default:
      updatedState = state;
    break;
  }

  if (process.env.NODE_ENV === `development`) {
    console.log(`***UPDATED_STATE***`, updatedState);
  }

  return updatedState;
}
