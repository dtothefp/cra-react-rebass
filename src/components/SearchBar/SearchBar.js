import React, { useState, useContext } from 'react';
import { Box, Button, Flex, Text } from 'rebass';
import { Input } from '@rebass/forms';
import { actions, StoreContext } from '../../store';

const SearchBar = () => {
  console.log('*****buttton', Button);
  const [state, setState] = useState('');
  const {dispatch} = useContext(StoreContext);
  const handleChange = (e) => {
    const {value} = e.target;

    console.log('change');

    setState(value);
  };

  const handleSubmit = () => {
    console.log('submitted');
  };

  return (
    <Flex as="form"
      onSubmit={handleSubmit}
      my={4}
    >
      <Input
        type="text"
        name="search"
        onChange={handleChange}
        placeholder="search"
        value={state}
        mr={2}
        width="70%"
      />
      <Button
        width="30%"
        color="black"
      >My Location</Button>
    </Flex>
  );
};

export default SearchBar;
