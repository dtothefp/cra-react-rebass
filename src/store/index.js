import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';
import reducer from './reducer';
import * as _actions from './actions';
import * as _constants from './constants';

export const actions = _actions;
export const constants = _constants;

export const StoreContext = createContext();

const _Provider = ({children, store = {}}) => {
  const [ state, dispatch ] = useReducer(reducer, store);

  return (
    <StoreContext.Provider value={{state, dispatch}}>
      {children}
    </StoreContext.Provider>
  );
};

_Provider.propTypes = {
  store: PropTypes.object,
  children: PropTypes.element.isRequired
}

export const Provider = _Provider;
