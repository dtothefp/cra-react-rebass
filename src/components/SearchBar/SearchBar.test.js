import React from 'react';
import {
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement
} from '@testing-library/react';
import * as Enzyme from 'enzyme';
import SearchBar from './SearchBar';
import Geocode from 'react-geocode';
import { StoreContext } from '../../store';
import mapsMock from '../../__mocks__/google.maps';

const fromAddress = jest.fn();

fromAddress.mockReturnValue({
  results: [
    {
      geometry: {
        location: {
          lat: `lat`,
          lng: `lng`,
        }
      }
    }
  ]
})

const fromLatLng = jest.fn();

fromLatLng.mockReturnValue({
  results: [
    {
      formatted_address: `address`,
    }
  ],
});

Object.assign(Geocode, {fromAddress, fromLatLng});

const MOCK_PHARMACY = `pharmacy`;
const MOCK_ADDRESS = `address`;
const MOCK_LAT = `lat`;
const MOCK_LNG = `lng`;

const fetchMock = jest.spyOn(global, `fetch`).mockReturnValue(
  Promise.resolve({
    json: () => ([
      MOCK_PHARMACY
    ])
  })
);

describe(`#SearchBar`, () => {
  const dispatch = jest.fn();
  const state = {filter: []};

  beforeAll(() => {
    class AutocompleteService {
      getPlacePredictions(input, cb) {
        cb([
          {
            description: MOCK_ADDRESS
          }
        ]);
      }
    }

    mapsMock.maps.places.AutocompleteService = AutocompleteService;
    global.google = mapsMock;

    const mockGeolocation = {
      getCurrentPosition: jest.fn()
        .mockImplementation((success) => Promise.resolve(success({
          coords: {
            latitude: MOCK_LAT,
            longitude: MOCK_LNG
          }
        })))
    };

    global.navigator.geolocation = mockGeolocation;
  });

  beforeEach(() => {
    fetchMock.mockClear();
  });

  it(`renders`, () => {
    expect(
      Enzyme.render(
        <StoreContext.Provider value={{dispatch, state}}>
          <SearchBar />
        </StoreContext.Provider>
      )
    ).toMatchSnapshot();
  });

  it(`fetches pharmacies when an auto-prediction is clicked`, async () => {
    const { getByText, getByRole, container } = render(
      <StoreContext.Provider value={{dispatch, state}}>
        <SearchBar />
      </StoreContext.Provider>
    );

    const input = getByRole(`textbox`);

    fireEvent.change(input, {target: {value: `data`}});

    await waitForDomChange({container});

    const li = await waitForElement(() => getByText(MOCK_ADDRESS), {container})

    fireEvent.click(li);

    await wait(() => {
      if (dispatch.mock.calls.length === 0) throw new Error(`dispatch not called`);
    });

    const url = fetchMock.mock.calls[0][0];

    expect(fetchMock.mock.calls.length).toBe(1)
    expect(url.endsWith(`location=${MOCK_LAT},${MOCK_LNG}`)).toBeTruthy();
  });

  it(`fetches pharmacies when the location button is clicked`, async () => {
    const { getByRole, container } = render(
      <StoreContext.Provider value={{dispatch, state}}>
        <SearchBar />
      </StoreContext.Provider>
    );

    const button = getByRole(`button`);

    fireEvent.click(button);

    await waitForDomChange({container});

    await wait(() => {
      if (dispatch.mock.calls.length === 0) throw new Error(`dispatch not called`);
    });

    const url = fetchMock.mock.calls[0][0];

    expect(fetchMock.mock.calls.length).toBe(1)
    expect(url.endsWith(`location=${MOCK_LAT},${MOCK_LNG}`)).toBeTruthy();
  });
});

