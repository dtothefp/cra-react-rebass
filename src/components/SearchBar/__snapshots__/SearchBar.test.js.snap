// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#SearchBar renders 1`] = `
<div>
  <div
    class="css-1qc3rwg"
  >
    <div
      class="css-ydkbfa"
    >
      <div
        class="css-1guzba0"
      >
        <input
          class="css-13hlwsa"
          placeholder="search"
          type="text"
          value=""
        />
        <ul
          class="css-1vuc2t3"
        />
      </div>
    </div>
    <button
      class="css-ngybz4"
    >
      My Location
    </button>
  </div>
</div>
`;
